compiler/rustc_mir_transform/src/ctfe_limit.rs

```diff
@@ -0,0 +1,58 @@
+//! A pass that inserts the `ConstEvalCounter` instruction into any blocks that have a back edge
+//! (thus indicating there is a loop in the CFG), or whose terminator is a function call.
+use crate::MirPass;
+
+use rustc_middle::mir::{
+    BasicBlock, BasicBlockData, BasicBlocks, Body, Statement, StatementKind, TerminatorKind,
+};
+use rustc_middle::ty::TyCtxt;
+
+pub struct CtfeLimit;
+
+impl<'tcx> MirPass<'tcx> for CtfeLimit {
+    #[instrument(skip(self, _tcx, body))]
+    fn run_pass(&self, _tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let indices: Vec<BasicBlock> = body
+            .basic_blocks
+            .iter_enumerated()
+            .filter_map(|(node, node_data)| {
+                if matches!(node_data.terminator().kind, TerminatorKind::Call { .. })
+                    // Back edges in a CFG indicate loops
+                    || has_back_edge(&body.basic_blocks, node, &node_data)
+                {
+                    Some(node)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        for index in indices {
+            insert_counter(
+                body.basic_blocks_mut()
+                    .get_mut(index)
+                    .expect("basic_blocks index {index} should exist"),
+            );
+        }
+    }
+}
+
+fn has_back_edge(
+    basic_blocks: &BasicBlocks<'_>,
+    node: BasicBlock,
+    node_data: &BasicBlockData<'_>,
+) -> bool {
+    let doms = basic_blocks.dominators();
+    basic_blocks.indices().any(|potential_dom| {
+        doms.is_reachable(potential_dom)
+            && doms.is_reachable(node)
+            && doms.is_dominated_by(node, potential_dom)
```

Comment:
It looks like iterating over `doms.dominators(node)` would reduce the complexity by the number of blocks factor.

Reply:
Thanks, I was able to improve the code a lot thanks to your comment 😊. Before, I tried to solve it this way because I couldn't figure out how to access the non-immediate dominators of each node, but now I see that the `doms.dominators(node)` method gives us an iterator over that data. I added some doc comments to `dominators/mod.rs` for later reference!